include/hint.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+namespace celerity {
+class handler;
+}
+
+// Definition is in handler.h to avoid circular dependency
+namespace celerity::experimental {
+template <typename Hint>
+void hint(handler& cgh, Hint&& h);
+}
+
+namespace celerity::detail {
+
+class hint_base {
+  public:
+	hint_base() = default;
+	hint_base(const hint_base&) = default;
+	hint_base& operator=(const hint_base&) = default;
+	hint_base(hint_base&&) = default;
+	hint_base& operator=(hint_base&&) = default;
+	virtual ~hint_base() = default;
+
+  private:
+	friend class celerity::handler;
+	virtual void validate(const hint_base& other) const {}
+};
+
+} // namespace celerity::detail
+
+namespace celerity::experimental::hints {
+
+/**
+ * Suggests that the task should be split into 1D chunks.
+ * This is currently the default behavior.
+ */
+class split_1d : public detail::hint_base {
+  private:
+	void validate(const hint_base& other) const override;
+};
+
+/**
+ * Suggests that the task should be split into 2D chunks.
+ */
+class split_2d : public detail::hint_base {
+  private:
+	void validate(const hint_base& other) const override;
+};
+
+inline void split_1d::validate(const hint_base& other) const {
+	if(dynamic_cast<const split_2d*>(&other) != nullptr) { throw std::runtime_error("Cannot combine split_1d and split_2d hints"); }
+}
+
+inline void split_2d::validate(const hint_base& other) const {
```

Comment:
Hmm true. The whole validation thing was a last minute addition after we decided to also add `split_1d`, before that there was no way of producing a conflict. The reason why I'm doing it here and not later is so I can throw directly within the call to `experimental::hint`. But since we wanted to pivot to aborting on these kinds of errors anyway, I guess it could also be deferred.

That being said, we may want to wait for another use case for more complex validation logic, `split_2d` also works for 1D and 0D kernels (the latter is a no-op).

Reply:
I would allow any hint in weird situations as long as it has reasonable fallback behavior. A `range<2>` can also be 1-dimensional at runtime.